File: java/src/main/java/com/smeltery/service/config/WebSecurityConfig.java
package com.smeltery.service.config;

import java.util.Arrays;
import java.util.Collections;

import com.smeltery.service.config.properties.Properties;
import com.smeltery.service.config.properties.WebSecurity;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.HeadersConfigurer;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;

@Configuration
@EnableWebSecurity
public class WebSecurityConfig {
	private final WebSecurity webSecurity;

	public WebSecurityConfig(Properties properties) {
		this.webSecurity = properties.getWebSecurity();
	}

	@Bean
	public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
		http
			.csrf()
			.disable()
			.cors()
			.configurationSource(corsConfigurationSource())
			.and()
			.headers()
			.frameOptions(HeadersConfigurer.FrameOptionsConfig::sameOrigin)
			.xssProtection((HeadersConfigurer<HttpSecurity>.XXssConfig xss) -> xss.xssProtectionEnabled(true));

		return http.build();
	}

	@Bean
	CorsConfigurationSource corsConfigurationSource() {
		CorsConfiguration configuration = new CorsConfiguration();
		configuration.setAllowedOrigins(webSecurity.getCorsAllowedOrigins());
		configuration.setAllowedMethods(Arrays.asList("GET", "POST", "HEAD", "PUT", "DELETE"));
		configuration.setAllowedHeaders(Collections.singletonList("*"));

		UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
		source.registerCorsConfiguration("/**", configuration);

		return source;
	}
}

File: kotlin/src/main/kotlin/com/ibm/service/config/WebSecurityConfig.kt
package com.smeltery.service.config

import com.smeltery.service.config.properties.Properties
import com.smeltery.service.config.properties.WebSecurity
import jakarta.servlet.http.HttpServletRequest
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configurers.HeadersConfigurer
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.header.writers.XXssProtectionHeaderWriter
import org.springframework.web.cors.CorsConfiguration
import java.util.*

@Configuration
@EnableWebSecurity
class WebSecurityConfig(properties: Properties) {

	private val webSecurity: WebSecurity = properties.webSecurity

	@Bean
	@Throws(Exception::class)
	fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
		http
			.csrf()
			.disable()
			.cors()
			.configurationSource { _: HttpServletRequest? ->
				val configuration = CorsConfiguration()
				configuration.allowedOrigins = webSecurity.getCorsAllowedOrigins()
				configuration.allowedMethods = mutableListOf("GET", "POST", "HEAD", "DELETE")
				configuration.allowedHeaders = listOf("*")
				configuration
			}
			.and()
			.headers()
			.frameOptions { obj: HeadersConfigurer<HttpSecurity>.FrameOptionsConfig -> obj.sameOrigin() }
			.xssProtection { xss: HeadersConfigurer<HttpSecurity>.XXssConfig ->
				xss.headerValue(
					XXssProtectionHeaderWriter.HeaderValue.ENABLED_MODE_BLOCK
				)
			}

		return http.build()
	}
}
